### PR TITLE
feat(pipelines): Docker registry credentials

### LINKED ISSUE #10999: [pipelines] Docker logins for assets
When using CDK Pipelines the autogenerated Assets action will build Docker images, and publish to the cdk-provided ECR. However if included Dockerfiles build on images in a non-public repository (e.g. an ECR in a different account), those builds will fail since the Assets action has no way of specifying sources to `docker login` in to.

### Use Case

Prior to using Pipelines we've used a shared ECR in a dedicated account to both store our internal base images, as well as images built on top of those. A single `docker login` would cover both pull and push from that repo.

However when switching to Pipelines, the destination repo changes. Pipelines transparently handles login to to that, but provides no configuration option for docker registries that need to be logged in to prior to asset building.

### Other

The error message from the Assets/DockerAsset1 CodeBuild project was:
```
Step 1/13 : FROM <account>.dkr.ecr.eu-west-1.amazonaws.com/...
| Get https://<account>.dkr.ecr.eu-west-1.amazonaws.com/...: no basic auth credentials
```

* [x] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### LINKED ISSUE #11774: (cdk-pipelines): Docker rate limit, guidance needed or auth mechanism
Currently can't deploy because of a rate limiting issue with Docker. You may have seen this in other issues, but the problem here is I have not hit this limit myself (only this single deploy) so for some reason, I believe Docker is flagging all of Codebuild as one "anonymous" source. 

Anyway, we need a way to authenticate when using `DockerImageAsset` from `@aws-cdk/aws-ecr-assets` or other guidance on how to circumvent this 

Error message: 

```bash
Step 1/9 : FROM node:14-alpine AS build
--
40 | toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit
41 | error  : [100%] fail: docker build --tag cdkasset-d0160342c2de47434404b7f88eaefd92e63e83f7178db2ff9a88cc4e0b62f5b2 --file ./docker/ci.Dockerfile . exited with error code 1: toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit
42 | Failure: Error: docker build --tag cdkasset-d0160342c2de47434404b7f88eaefd92e63e83f7178db2ff9a88cc4e0b62f5b2 --file ./docker/ci.Dockerfile . exited with error code 1: toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit
```

### Proposed Solution

We need a way to authenticate to associate our accounts with the pull instead of anon.

### PR DESCRIPTION
**Pull request by njlynch**
Introduce a new set of properties to the pipeline constructs to enable users to
specify Docker registries -- and associated credentials for each -- to be used
during the pipeline build/synth, self-mutate, and asset publishing stages.

These APIs enable the user to specify a Docker registry (e.g., DockerHub, ECR)
and either secrets or use role credentials to authenticate to each registry, as
well as specify which step(s) of the pipeline need these credentials.

fixes #10999
fixes #11774

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/pipelines/lib/docker-registry.ts [hunk 1]
```diff
@@ -0,0 +1,164 @@
+import * as ec2 from '@aws-cdk/aws-ec2';
+import * as ecr from '@aws-cdk/aws-ecr';
+import * as iam from '@aws-cdk/aws-iam';
+import * as secretsmanager from '@aws-cdk/aws-secretsmanager';
+import { Fn } from '@aws-cdk/core';
+
+/** Docstrings */
+export abstract class DockerRegistry {
+  /** Docstrings */
+  public static fromDockerHub(opts: ExternalDockerRegistryOptions): DockerRegistry {
```
Comment (rix0rrr) [inline packages/@aws-cdk/pipelines/lib/docker-registry.ts:10]: To my mind, `from` is the wrong word here. It's either `forDockerHub`, or just `dockerHub` I'd say.

You're not making a `DockerRegistry`: you're describing one. Or the class should be called `DockerCredentials`...

File: packages/@aws-cdk/pipelines/lib/docker-registry.ts [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+   * Docstrings
+   * @internal
+   */
+  public abstract _renderCdkAssetsConfig(): DockerRegistryCredentialSource
+}
+
+/** Docstrings */
+export interface ExternalDockerRegistryOptions {
+  /** Docstrings */
+  readonly secret: secretsmanager.ISecret;
```
Comment (rix0rrr) [inline packages/@aws-cdk/pipelines/lib/docker-registry.ts:38]: Would it make sense to make `secret` a positional argument? Or is that crazy?

File: packages/@aws-cdk/pipelines/lib/docker-registry.ts [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+  constructor(private readonly repositories: ecr.IRepository[], private readonly opts: EcrDockerRegistryOptions) {
+    super();
+
+    if (repositories.length === 0) {
+      throw new Error('must supply at least one `ecr.IRepository` to create an `EcrDockerRegistry`');
+    }
+    this.registryDomain = Fn.select(0, Fn.split('/', repositories[0].repositoryUri));
+  }
+
+  public grantRead(grantee: iam.IGrantable) {
```
Comment (rix0rrr) [inline packages/@aws-cdk/pipelines/lib/docker-registry.ts:108]: Needs the usage as a parameter, no?

File: packages/@aws-cdk/pipelines/lib/docker-registry.ts [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+  readonly secretsPasswordField?: string;
+  readonly ecrRepository?: boolean;
+  readonly assumeRoleArn?: string;
+}
+
+export function dockerRegistriesInstallCommands(registries?: DockerRegistry[], osType?: ec2.OperatingSystemType): string[] {
+  if (!registries || registries.length === 0) { return []; }
+
+  const domainCredentials = registries.reduce(function(map: Record<string, any>, registry) {
+    map[registry.registryDomain] = registry._renderCdkAssetsConfig();
```
Comment (rix0rrr) [inline packages/@aws-cdk/pipelines/lib/docker-registry.ts:143]: We could do without making `registryDomain` public if we did:

```ts
public _renderCdkAssetConfig() {
  return { [this.registryDomain]: { ...whatevs ... }};
}

// and here
Object.assign(map, registry._renderCdkAssetsConfig());
```

Just a thought

File: packages/@aws-cdk/pipelines/lib/docker-registry.ts [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+    super();
+
+    if (repositories.length === 0) {
+      throw new Error('must supply at least one `ecr.IRepository` to create an `EcrDockerRegistry`');
+    }
+    this.registryDomain = Fn.select(0, Fn.split('/', repositories[0].repositoryUri));
+  }
+
+  public grantRead(grantee: iam.IGrantable) {
+    if (this.opts.assumeRole) {
```
Comment (rix0rrr) [inline packages/@aws-cdk/pipelines/lib/docker-registry.ts:109]: Why don't we have `Role#grantAssumeRole()` ??

Comment (njlynch) [reply packages/@aws-cdk/pipelines/lib/docker-registry.ts:109]: Good question...

File: packages/@aws-cdk/pipelines/lib/docker-registry.ts (hunk 4 above)
Comment (njlynch) [reply packages/@aws-cdk/pipelines/lib/docker-registry.ts:143]: I like it. I was trying to do something similar, but couldn't get the syntax just right.

**Review by rix0rrr — APPROVED** [review]

**Review by rix0rrr — APPROVED** [review]

**Comment by adriantaut** [comment]
@njlynch is this PR supposed to work with `ghcr.io` or with `docker.pkg.github.com`? simply can't make it work

**Pull request merged**